File: day2/solution.ts
import { readLineAsArray } from "../common/io";

const INPUT_FILEPATH = "day2/input.txt";
const input = readLineAsArray(INPUT_FILEPATH);

const parsedInput = input.reduce((acc, line) => {
  const [gameName, gamesData] = line.split(":");
  const gameId = gameName.split(" ")[1];

  const rawDrawList = gamesData.split(";");
  const drawList = rawDrawList.map((rawDraw) => {
    const rawItems = rawDraw.split(",");

    const items = rawItems.reduce((acc, rawItem) => {
      rawItem = rawItem.trim();

      const [quantity, item] = rawItem.split(" ");
      acc[item] = +quantity;
      return acc;
    }, {});

    return items;
  });

  acc[gameId] = drawList;
  return acc;
}, {});


const partOne = () => {
  const check = {
    red: 12,
    green: 13,
    blue: 14,
  };
  const validIds = [];
  for (let gIdx in parsedInput) {
    let isInvalid = false;
    const draws = parsedInput[gIdx];
    for (let draw of draws) {
      for (let key in draw) {
        if (draw[key] > check[key]) {
          isInvalid = true;
          break;
        }
      }
    }
    if (!isInvalid) {
      validIds.push(+gIdx);
    }
  }

  const sum = validIds.reduce((acc, item) => acc + item)
  console.log("PartOne", sum);
};
partOne();

const partTwo = () => {
    let result = 0;
    for (let gIdx in parsedInput) {
        const maxBalls = {
            red: 0,
            green: 0,
            blue: 0
        }
        const draws = parsedInput[gIdx]
        for (let draw of draws) {
            for (let key in draw) {
                if (draw[key] > maxBalls[key]) {
                    maxBalls[key] = draw[key]
                }
            }
        }

        const power = Object.values(maxBalls).reduce((acc, val) => acc * val, 1)
        result += power
    }

    console.log("PartTwo", result)
}
partTwo()

File: day2/solution.py
import sys
import os
from functools import reduce

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
from common.io import readLineAsArray

INPUT_FILEPATH = "day2/input.txt"
input = readLineAsArray(INPUT_FILEPATH)

def raw_items_reducer(acc, item):
    rawItem = item.strip()

    quantity, item = rawItem.split(" ")
    acc[item] = int(quantity)
    return acc

def parsed_input_reducer(acc, line):
    gameName, gamesData = line.split(":")
    gameId = gameName.split(" ")[1]

    rawDrawList = gamesData.split(";")
    drawList = []
    for draw in rawDrawList:
        rawItems = draw.split(",")
        items = reduce(raw_items_reducer, rawItems, {})
        drawList.append(items)
    
    acc[gameId] = drawList
    return acc

# Parse the Input
parsed_input = reduce(parsed_input_reducer, input, {})

def part_one():
    check = {
        "red": 12,
        "green": 13,
        "blue": 14
    }
    possibleGames = []
    for gameId,gameData in parsed_input.items():
        possible = True
        for draw in gameData:
            if (draw.get("red") != None and draw["red"] > check["red"]):
                possible = False
                break
            if (draw.get("blue") != None and draw["blue"] > check["blue"]):
                possible = False
                break
            if (draw.get("green") != None and draw["green"] > check["green"]):
                possible = False
                break
        if possible == True:
            possibleGames.append(int(gameId))

    sum = 0
    for gameId in possibleGames:
        sum += gameId
    print("PartOne", sum)

def part_two():
    result = 0
    for _, gameData in parsed_input.items():
        min_cubes = {
            "red": 0,
            "green": 0,
            "blue": 0
        }

        for draw in gameData:
            for key in draw:
                if draw[key] > min_cubes[key]:
                    min_cubes[key] = draw[key]

        power = 1
        for _, val in min_cubes.items():
            power *= val

        result += power
    
    print("PartTwo", result)


part_one()
part_two()